support-workers/src/main/scala/com/gu/support/workers/lambdas/GetPaymentMethod.scala

```diff
@@ -0,0 +1,83 @@
+import com.gu.support.zuora.api.response.{GetPaymentMethodCardReferenceResponse, GetPaymentMethodResponse}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class GetPaymentMethod(servicesProvider: ServiceProvider = ServiceProvider)
+    extends ServicesHandler[GetPaymentMethodState, CreateZuoraSubscriptionState](servicesProvider) {
+
+  def this() = this(ServiceProvider)
+
+  override protected def servicesHandler(state: GetPaymentMethodState, requestInfo: RequestInfo, context: Context, services: Services) = {
+
+    val zuoraService = services.zuoraService
+    for {
+      account <- zuoraService.getObjectAccount(state.paymentFields.billingAccountId)
+      accountIdentityId <- getOrFailWithMessage(account.IdentityId__c, "Zuora account has no identityId")
+      _ <- ifFalseReturnError(accountIdentityId == state.user.id, s"Zuora identity id: $accountIdentityId does not match user identity id ${state.user.id}")
+      paymentId <- getOrFailWithMessage(account.DefaultPaymentMethodId, "Zuora account has no default payment method")
+      getPaymentMethodResponse <- zuoraService.getPaymentMethod(paymentId)
+      _ <- ifFalseReturnError(getPaymentMethodResponse.paymentMethodStatus == "Active", "Zuora account has a non active default payment method")
+      sfContactId <- getOrFailWithMessage(account.sfContactId__c, "Zuora account has no sfContact")
+      paymentMethod <- asFuture(toPaymentMethod(getPaymentMethodResponse))
+      sfContact = SalesforceContactRecord(sfContactId, account.CrmId)
+    } yield  HandlerResult(
+        CreateZuoraSubscriptionState(
+          requestId = UUID.randomUUID(),
+          user = state.user,
+          product = state.product,
+          paymentMethod =  paymentMethod,
+          firstDeliveryDate = None,
+          promoCode = None,
+          salesForceContact = sfContact,
+          acquisitionData = state.acquisitionData
+        ),
+      requestInfo
+
+    )
+  }
+
+  def toPaymentMethod(getPaymentMethodResponse: GetPaymentMethodResponse): Either[String, PaymentMethod] = getPaymentMethodResponse match {
+
+    case cardResponse: GetPaymentMethodCardReferenceResponse => {
+      val maybeCountry: Option[Country] = cardResponse.creditCardCountry.flatMap(CountryGroup.byOptimisticCountryNameOrCode(_))
+      Right(
+        CreditCardReferenceTransaction(
+          tokenId = cardResponse.tokenId,
+          secondTokenId = cardResponse.secondTokenId,
+          creditCardNumber = cardResponse.creditCardMaskNumber,
+          creditCardCountry = maybeCountry,
+          creditCardExpirationMonth = cardResponse.creditCardExpirationMonth,
+          creditCardExpirationYear = cardResponse.creditCardExpirationYear,
+          creditCardType = cardResponse.creditCardType
+        )
+      )
+    }
+    case other => Left(s"unsupported payment method: '$other'")
+
+  }
+
+  def errorResponse(msg:String): Future[Nothing] = Future.failed(new Exception(msg))
```

Comment:
At the moment an unhandled exception will cause the Step Function execution to fail and result in an alert. If any of these errors are expected we should handle them gracefully here: https://github.com/guardian/support-frontend/blob/master/support-workers/src/main/scala/com/gu/support/workers/lambdas/FailureHandler.scala#L49-L68

Reply:
@twrichards  might know more about this, but as far as I understand frontend will only offer to reuse payment methods in cases where all these validations pass.

Reply:
Correct and so we would want alerts as it would indicate something is mis-implemented or being used maliciously

Reply:
Also, I just realised that a lot of this validations are not needed. I was initially replicating the csr acquisitions code where we need to return a nice error message to the user. But in this case i can make the fields non optional in the case class and let the json parsing fail if something required is missing

Reply:
.. Unless we want to leave a more generic "get zuora account" function in support workers for future reuse
🤔